plutus-ir/src/Language/PlutusIR/Transform/LetFloat.hs

```diff
@@ -0,0 +1,455 @@
+                                         Just rhs ->
+                                           -- lift each resulting binding in a list monoid to accumulate them
+                                           over _3 pure $
+                                           -- visit the generated rhs-term as well for any potential floating
+                                           over (_3.bindingSubterms)
+                                                (goTerm curDepth restDepthTable)
+                                                rhs
+                                         _ -> error "Something went wrong: no rhs was found for this let in the rhstable."
+                                   ) vs
+                     in case acc of
+                       -- merge current let-group with previous let-group iff both groups' recursivity is NonRec
+                       Let accAnn NonRec accBs accIn | mRecurs == NonRec -> Let (mAnn <> accAnn) NonRec (newBindings <> accBs) accIn
+                       -- never merge if the previous let-group is a Rec or the current let-group is Rec,
+                       -- but instead create a nested current let-group under the previous let-group (above)
+                       _ -> Let mAnn mRecurs newBindings acc
+               else acc -- skip this scc, we are not interested in it right now
+             ) t topSortedSccs
+
+
+-- | The main transformation function (Term -> Term) that returns a term with all lets floated as outwards as possible (full laziness).
+-- Traverses the AST:
+-- 1stpass: to collect the maximum ranks (positions) of all lets
+-- 2ndpass:  to remove all its lets and place them back (float them) in their maximum ranks (positions).
+--
+-- Note: this transformation requires that the PLC.rename compiler-pass has prior been run.
+floatTerm :: forall name tyname uni a
+           . (Ord (tyname a), PLC.HasUnique (tyname a) PLC.TypeUnique, PLC.HasUnique (name a) PLC.TermUnique, Monoid a)
+          => Term tyname name uni a -> Term tyname name uni a
+floatTerm pir = p2Term pir
+                -- give the floatdata of the 1st pass to the start of the 2nd pass
+              $ p1Term pir
+
+
+-- Helpers
+----------
+
+-- | Returns a single 'Unique' for a particular binding.
+-- We need this because let-datatypes introduce multiple identifiers, but in our 'RhsTable', we use a single Unique as the key. See also: 'bindingIds'
+principal :: (PLC.HasUnique (tyname a) PLC.TypeUnique, PLC.HasUnique (name a) PLC.TermUnique)
+            => Getting r (Binding tyname name uni a) PLC.Unique
+principal = to $ \case TermBind _ _ (VarDecl _ n _) _ -> n ^. PLC.unique . coerced
+                       TypeBind _ (TyVarDecl _ n _) _ -> n ^. PLC.unique . coerced
+                       -- arbitrary: uses the match-function's unique as the principal unique of this data binding group
+                       DatatypeBind _ (Datatype _ _ _ matchFunc _) -> matchFunc ^. PLC.unique . coerced
```

Comment:
Nitpicking, but I'd choose the `TyVarDecl`'s unique rather than the matching function's one, because in the two other cases we choose the unique of the thing being defined, which in this case is the name of a data type. It's also the case that the unique of the name of a data type is the outermost unique of a data type declaration, so I think it makes more sense to make this unique represent the whole declaration than to pick some arbitrary unique from the middle of the declaration.